Block list: Add option in editor settings to show open list view as default

ramonjd opened the pull request:
Resolves https://github.com/WordPress/gutenberg/issues/40565


## What?

A preference to have the block list view open by default. 🪄 

https://user-images.githubusercontent.com/6458278/166365673-57cda91b-26f4-45cf-8c6a-d897dd8a1567.mp4

## Why?

Some folks use the block list view all the time. 

Creating an option to have it opened by default will save these folks approximately a couple of seconds every time the editor loads. Over a year this could save up to forty-two minutes.

## How?

Adding an extra feature option to the post and site editor preferences modals.

**Note**: the post editor preferences modal has not yet been migrated over to the preference store, hence the use of `isFeatureActive` in [packages/edit-post/src/components/layout/index.js](https://github.com/WordPress/gutenberg/pull/40757/files#diff-8933c0b377d83066f7dde9c8b573bab9a7bfe60f73919e4e5e80aff92a88800cR134) 

See site editor migration over at https://github.com/WordPress/gutenberg/pull/39485


## Testing Instructions

1. Open up the post editor. Notice the block list sidebar is not open.
2. Open the preferences modal and turn on  `Always open list view` 
3. Refresh page.
4. Wait a bit.
5. Check that the block list sidebar is open.
6. Think about what you're having for dinner. 
7. Repeat steps again for the site editor.
8. Be well.

Run the e2e playwright test for glory! 

```bash
npm run test-e2e:playwright -- site-editor/block-list-panel-preference.spec.js
```

ramonjd committed:
initial commit

ramonjd committed:
Added option to site editor

ramonjd committed:
Change key to `showListViewByDefault`
Add basic e2e test
Update snapshots
Add example single test run to playwright readme

ramonjd commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -129,6 +131,9 @@ function Layout( { styles } ) {
 			),
 			// translators: Default label for the Document in the Block Breadcrumb.
 			documentLabel: postTypeLabel || _x( 'Document', 'noun' ),
+			isListViewOpenByDefault: select( editPostStore ).isFeatureActive(
```
@talldan 

Do you know if the post editor preferences modal is, or should be, migrated over to the preferences store?  Done for the site editor in https://github.com/WordPress/gutenberg/pull/39485

I'm happy to take a look if you think it's worth it.

talldan replied:
@ramonjd Do you mean the component or the preferences data?

The situation is:
- There's a generic preferences modal component in the interface package and that's used in the post and site editors.
- All preferences data is in the preferences package for those editors (`select( editPostStore ).isFeatureActive` is a registry selector that calls through to the preferences store).

So I think you're all good. At some point I'll look at deprecating the post editor's `isFeatureActive` selector.

ramonjd replied:
> Do you mean the component or the preferences data?

Ah I was looking at the [EnableFeature component](https://github.com/WordPress/gutenberg/blob/d10329149ce45ef3dd9572fb0a03ebff548cc092/packages/edit-post/src/components/preferences-modal/options/enable-feature.js). It uses `isFeatureActive`.

I'm just comparing it with the corresponding [EnableFeature component](https://github.com/WordPress/gutenberg/blob/8d0a84f7edaa090acb26d7d30c500ace30da0a45/packages/edit-site/src/components/preferences-modal/enable-feature.js) for the site editor preferences, which uses the preferences store.

> (select( editPostStore ).isFeatureActive is a registry selector that calls through to the preferences store)

Great, thanks for confirming. 

I was mainly checking to see if how I'm selecting the value of the preference for the [post](https://github.com/WordPress/gutenberg/pull/40757/files#diff-8933c0b377d83066f7dde9c8b573bab9a7bfe60f73919e4e5e80aff92a88800cR134) and [site](https://github.com/WordPress/gutenberg/pull/40757/files#diff-53f39da7a5f1f996f5359a2dd7b2747fb3d60a494523dd73dba946fab8e82592R118) editors is okay for now. 

I mean, it works 😄

ramonjd commented on test/e2e/specs/site-editor/block-list-panel-preference.spec.js:
```diff
@@ -0,0 +1,40 @@
+		await expect(
+			page.locator( '.edit-site-editor__list-view-panel' )
+		).toBeVisible();
+	} );
```
@kevin940726 to check that my playwright test isn't a total mess 😄 

I couldn't get something like `'role=region[aria-label="List View"]'` working, hence the classname

talldan replied:
How about something like `'role=region[name="List View"i]'`?

(The `i` means case insensitive. The `name` is part of the role selector that playwright has - https://playwright.dev/docs/selectors#role-selector)

ramonjd replied:
> How about something like 'role=region[name="List View"i]'?

Thanks, that works! 🍺

I was naively trying with `aria-label`

ramonjd committed:
Use `name` in locator selector

glendaviesnz commented on packages/edit-post/src/components/preferences-modal/test/__snapshots__/index.js.snap:
```diff
@@ -32,6 +32,11 @@ exports[`EditPostPreferencesModal should match snapshot when the modal is active
                 help="Highlights the current block and fades other content."
```
Are these snapshot tests providing any value - I thought at some point there was a move to stop adding snapshot tests?

glendaviesnz replied:
Here is the note I was thinking about in relation to snapshots https://github.com/WordPress/gutenberg/pull/32592

ramonjd replied:
> Are these snapshot tests providing any value

Hmmm I'm not sure in this case. 🤔 

They're only part of this PR because they're pre-existing and I updated them with the changes.

Having said that, if you think they have little value in this context I can remove them in a new PR.

glendaviesnz replied:
I guess in the case of these snapshot tests they are doing something other than just checking that the component today looks exactly the same as it did yesterday - it is checking if it adapts correctly for given screen sizes, which would take a bit of additional work to test from a frontend functionality perspective ... so I would be inclined to leave it as-is for now given it is an area of code that doesn't get touched often.

ramonjd replied:
Thanks for taking the time to confirm @glendaviesnz !

glendaviesnz commented:
This tested as advertised for me - even across browsers. Just left a question about the value of the snapshot test, but that doesn't need to be a blocker to getting this merged I don't think.

kevin940726 commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -142,6 +147,13 @@ function Layout( { styles } ) {
+		if ( isListViewOpenByDefault ) {
+			setIsListViewOpened( true );
+		}
+	}, [] );
```
I think we should start respecting the [exhaustive deps](https://github.com/facebook/react/issues/14920#issuecomment-471070149) lint rule. There's an open PR and discussion in https://github.com/WordPress/gutenberg/pull/24914. (tldr; while it'll introduce some false positives, I still think it's worth enabling it in the long run.)

As for this example, both `setIsListViewOpened` and `isListViewOpenByDefault` should be included. If we don't want to react to changes for `isListViewOpenByDefault`, we could wrap it in a ref to make it explicit.

```js
const isListViewOpenByDefaultRef = useRef( isListViewOpen );

useEffect( () => {
	if ( isListViewOpenByDefaultRef.current ) {
		setIsListViewOpened( true );
	}
}, [ setIsListViewOpened ] );
```

(BTW, shouldn't it make more sense to rename the variable `shouldListViewOpenByDefault`? 😆)

kevin940726 replied:
On the other hand, I think we could dispatch the action immediately before rendering the React tree to update the state to prevent it from unnecessary re-rendering. IIRC, we do this in the site editor.

Inside `initializeEditor` (or `reinitializeEditor`, not sure), we could dispatch action before starting the render.

```js
function initializeEditor() {
	if ( select( editPostStore ).isFeatureActive( 'showListViewByDefault' ) ) {
		dispatch( editPostStore ).setIsListViewOpened( true );
	}

	render( // ...
}
```

ramonjd replied:
Thanks for the detailed notes @kevin940726 🙇 

I just wanted things to run on mount, your `initializeEditor` suggestion sounds promising. I'll look into it.

kevin940726 commented on packages/edit-site/src/components/editor/index.js:
```diff
@@ -152,6 +159,13 @@ function Editor( { onError } ) {
 	);
 
+	// Check if the block list view should be open by default.
+	useEffect( () => {
```
Ditto above.

ramonjd committed:
Move dispatch to initialize editor methods to avoid unnecessary re-renders using useEffect

andrewserong commented on packages/edit-post/src/index.js:
```diff
@@ -123,6 +123,12 @@ export function initializeEditor(
 	} );
```
For completeness sake, should we add a `showListViewByDefault: false` line to the above `setDefaults` to match the one you've added in `edit-site`?

ramonjd replied:
Yes, totally!

I think I left it out because I wasn't sure if the key was the same in the preferences store (as opposed to the older features store), but it seems to have been synched.

Anyway, I'll add it. Good one!

andrewserong reviewed (approved):
This works great for me, I think it'll be a really useful feature for folks who prefer to design their posts and sites via the list view 👍

Separately / in a follow-up as the site editor gets more features added to the preferences modal, we might want to look at the sorting of the feature and where it appears in the list, as the lists aren't quite consistent at the moment (`Always open list view` is third in the list in both cases, but the `Display button labels` is in a different position and has a different label):

| Post editor | Site editor |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/167544435-d2afa9ac-1176-47b6-a60c-86227ff408ef.png) | ![image](https://user-images.githubusercontent.com/14988353/167544455-19a2b3e0-c4de-40a7-9214-31e96fccc9d2.png) |

Keep this feature 3rd in the list in both places makes it easy to find though, so I don't think we need to change anything just yet.

I also added a comment about a not-blocking nit of whether we should explicitly set the default in `edit-post` like we've got the default in `edit-site`?

But no blockers from me, I think this looks great! 🎉

ramonjd committed:
Set default value of showListViewByDefault to `false` in the post editor.

ramonjd commented:
Thanks for testing this little one @andrewserong !

> Separately / in a follow-up as the site editor gets more features added to the preferences modal, we might want to look at the sorting of the feature and where it appears in the list, as the lists aren't quite consistent at the moment (Always open list view is third in the list in both cases, but the Display button labels is in a different position and has a different label):

Excellent point. I think they should be consistent in both copy and order. I'll add a follow up to this PR after merging.

>  also added a comment about a not-blocking nit of whether we should explicitly set the default in edit-post like we've got the default in edit-site?

Fixed! 🙇

kevin940726 reviewed (approved):
This looks good to me 👍 !

paaljoachim commented:
Lately I noticed that List View kept opening. I went to a page List View was open. I closed it. I exited the page and came back to it later on. I noticed again that List View was open. I closed it again. I viewed the page and came back to it. List View was open again. I thought it was a bug.

I wrote about this in the core editor channel on Slack.
@ramonjd replied and mentioned this PR.

In general the user would not discover the option has been added to Preferences, and will probably do as I was, click to close List View over and over again. There should be a kind of remembrance. When the user closes List View is should be closed until the user decides to reopen it. Or else the program decides for the user giving the user less control over the program. As a user I want to control the program. I want the program to remember when I open and close something. That it does not reopen something I have already closed.

As the new option has been added to preferences it should by default be off. It is better for the user to discover the option because they want something to happen than go to preferences because they want to turn something off.

Often I do not use List View because I just go to a post/page and make simple adjustments. List View would be nice to use for more complex adjustments. When I visit various web sites I handle there are most of the times just simple adjustments to make where I do not need to use List View.

ramonjd commented:
> I was the new option has been added to preferences it should by default be off.

Interesting. The intention was, and testing back then demonstrated, that the default was `false` (off).

See the [site editor](https://github.com/WordPress/gutenberg/blob/27ec8f65c929e7e96ecb73aa1097fad5f55e0d1c/packages/edit-site/src/index.js#L68) and [post editor](https://github.com/WordPress/gutenberg/blob/f60ca37718337f64b9a446b684c81d3ab4f07d4d/packages/edit-post/src/index.js#L120) default preferences. 

I can take a quick look in the morning to see if there's a bug there or something has changed in the meantime.

paaljoachim commented:
Great it is a bug I experienced!
By default it should be off. When it is actually is on by default.

ramonjd commented:
@paaljoachim Thanks for the info here!

Quick question: does the list view persist if you clear local storage? `localStorage.clear()`

If the list view doesn't show up on refresh, then things are working as intended. I tested this and it looks okay.

Another way to test this while developing is to run `localStorage.clear()`, refresh the post editor. Check that the block list is not displayed. Then manually edit [the showListViewByDefault preference default value](https://github.com/WordPress/gutenberg/blob/f60ca37718337f64b9a446b684c81d3ab4f07d4d/packages/edit-post/src/index.js#L120) to `true`.

Refresh the post editor and it should show by default (nothing will appear in local storage until you manually toggle the preferences using the preferences modal).

If this is not working for you, then let me know the steps you took to recreate the issue and I'll dig deeper. 

On a side note I found a small typo in the site editor default preferences, so thanks for bringing this issue to my attention!

This would have had no effect on whether the block list displayed by default or not. I threw up a PR: 

- https://github.com/WordPress/gutenberg/pull/43494

ramonjd commented:
Also, @talldan reminded me that preferences are stored in user meta in the database, so you also might need to remove those to test, or, if you're using the wp-env environment, destroy/rebuild.

```js
wp.apiFetch( {
	path: '/wp/v2/users/me',
	method: 'PUT',
	data: {
		meta: {
			persisted_preferences: {},
		},
	},
} );
```

talldan commented:
I double-checked this, as I had recently set up a fresh local WordPress environment. The preference definitely defaulted to switched off for me.

@paaljoachim Is there a chance you may have turned it on accidentally? Maybe try testing using something like Local by Flywheel so that you have a completely fresh instance of WordPress.

paaljoachim commented:
Hi Dan and Ramon.

I met this List View option through a full site editing local web site that I have. I use it for testing various Gutenberg PR builds, and general exploration. For whatever reason List View began to open up every time I entered a page or post, even with the regular Gutenberg plugin installed. I deactivated and removed and installed the Gutenberg plugin again. Then turned off the preference option. 

I have tested other local sites that I have and also a few online, and did not see the same thing there.
I will just call it a glitch in the system and move on.

Thank you! Calling this solved. Gremlin removed from the test site...:)
